File: src/main/java/com/grudus/adventofcode/Day08.java
package com.grudus.adventofcode;

import io.vavr.Function2;
import io.vavr.Tuple2;
import io.vavr.collection.List;
import io.vavr.collection.Map;
import io.vavr.collection.Stream;

import java.util.HashSet;
import java.util.Set;

public class Day08 extends Day<Integer> {
    record State(Integer lineNumber, Integer value) {}

    record Instruction(InstructionType type, Integer value) {
        public State execute(State state) {
            return type.action.apply(state, value);
        }
        public boolean canToggle() {
            return type != InstructionType.ACC;
        }
        public Instruction toggle() {
            return type == InstructionType.JMP
                    ? new Instruction(InstructionType.NOP, value)
                    : new Instruction(InstructionType.JMP, value);
        }
    }

    enum InstructionType {
        ACC((state, val) -> new State(state.lineNumber + 1, state.value + val)),
        JMP((state, val) -> new State(state.lineNumber + val, state.value)),
        NOP((state, val) -> new State(state.lineNumber + 1, state.value));
        public final Function2<State, Integer, State> action;
        InstructionType(Function2<State, Integer, State> action) {
            this.action = action;
        }
    }

    @Override
    public Integer firstStar(List<String> input) {
        Map<Integer, Instruction> instructions = input
                .map(this::parseInstruction)
                .zipWithIndex()
                .toMap(Tuple2::_2, Tuple2::_1);

        return runInstructions(instructions).value;
    }

    @Override
    public Integer secondStar(List<String> input) {
        Map<Integer, Instruction> instructions = input
                .map(this::parseInstruction)
                .zipWithIndex()
                .toMap(Tuple2::_2, Tuple2::_1);

        return Stream.range(0, instructions.size())
                .filter(i -> instructions.get(i).get().canToggle())
                .map(i -> instructions.put(i, instructions.get(i).get().toggle()))
                .map(this::runInstructions)
                .find(state -> state.lineNumber >= instructions.size())
                .get().value;
    }

    private State runInstructions(Map<Integer, Instruction> instructions) {
        Set<Integer> executedLines = new HashSet<>();

        return Stream.iterate(new State(0, 0),
                state -> instructions.getOrElse(state.lineNumber, new Instruction(InstructionType.JMP, 1)).execute(state))
                .takeWhile(state -> state.lineNumber <= instructions.size())
                .takeWhile(state -> !executedLines.contains(state.lineNumber))
                .peek(state -> executedLines.add(state.lineNumber))
                .last();
    }

    private Instruction parseInstruction(String s) {
        return new Instruction(
                InstructionType.valueOf(s.substring(0, 3).toUpperCase()),
                Integer.parseInt(s.substring(4))
        );
    }
}

File: src/main/python/day08.py
def first_star(instructions):
    return execute_code(list(map(parse_instruction, instructions)))[1]
    

def second_star(instructions):
    instructions = list(map(parse_instruction, instructions))

    for i, (operation, value) in enumerate(instructions):
        new_instructions = instructions.copy()

        if operation == 'jmp':
            new_instructions[i] = ('nop', value)
        elif operation == 'nop':
            new_instructions[i] = ('jmp', value)
        
        (execution_type, state) = execute_code(new_instructions)
        if execution_type == 'END':
            return state


def execute_code(parsed_instructions):
    global_state = 0
    current_line = 0
    executed_lines = set()
    all_lines = len(parsed_instructions)
    
    while True:
        (operation, value) = parsed_instructions[current_line]

        if operation == 'nop':
            current_line += 1
        elif operation == 'acc':
            current_line += 1
            global_state += value
        elif operation == 'jmp':
            current_line += value

        if (current_line in executed_lines):
            return ("LOOP", global_state)
        elif (current_line >= all_lines):
            return ("END", global_state)

        executed_lines.add(current_line)


def parse_instruction(instruction):
    operation = instruction[:3]
    value = instruction[4:]
    return (operation, int(value[1:]) if value[0] == '+' else -int(value[1:]))

if __name__ == "__main__":
    instructions = open('src/main/resources/day08/input.txt', 'r').read().split("\n")
    print(first_star(instructions))
    print(second_star(instructions))
